generate_directory_indexes.py

```diff
@@ -115,47 +156,131 @@ def format_directory_listing(directory_listing):
+                    last_modified_format)
+                size = int(metadata[3])
+
+            file_details.append(
+                file_information(full_path, file_name, last_modified, size))
+
+    return file_details
+
+def gather_file_details(current_path, list_of_files):
+    file_details = []
     # generate indexes for current path
-    for file in contents:
+    for file_name in list_of_files:
         # add size, lastModified, file/folder type to metadata
-        fullpath = os.path.join(path, file)
-        if bool(re.match('.*index_by.*\.html$', file)):
+        full_path = os.path.join(current_path, file_name)
+        if is_an_index_file(file_name):
             continue
-        if(os.path.exists(fullpath)):
-            icon = 'folder.gif' if os.path.isdir(fullpath) else 'unknown.gif'
-            directory_listing.append({
-                'name': file,
-                'lastModified': os.path.getmtime(fullpath),
-                'size': os.path.getsize(fullpath),
-                'icon': icon
-            })
+        if os.path.exists(full_path):
+            icon = 'folder.gif' if os.path.isdir(full_path) else 'unknown.gif'
+            file_details.append(file_information(full_path, file_name))
         else:
-            logging.error('skipping \'{}\' because the file cannot be read'.format(fullpath))
+            logging.error('skipping \'{}\' because the file cannot be read'.format(full_path))
+
+    return file_details
+
+def make_index_files(base_path, current_path, file_details):
     for order_by in ['name', 'size', 'lastModified']:
         for reverse_order in [True, False]:
-            file_name = os.path.join(path, index_file_name('', order_by, reverse_order))
-            rendered_html = render_index(path, order_by, directory_listing, reverse_order, base_path)
-            if args.noop:
-                logging.info('Would create: ', file_name)
+            file_name = os.path.join(current_path, index_file_name('', order_by, reverse_order))
+            rendered_html = render_index(current_path, order_by, file_details,
+                                         reverse_order, base_path)
+            if configuration.noop:
+                logging.info('Would create: {}'.format(file_name))
             else:
                 logging.info('Wrote: {}'.format(file_name))
                 index_file = open(file_name, 'w')
                 index_file.write(rendered_html)
                 index_file.close()
 
 
-    # recurse into subdirectories
-    for file in contents:
-        file_pwd = os.path.join(path, file)
-        if os.path.isdir(file_pwd):
-            walk(file_pwd, base_path)
+def traverse_tree(base_path, current_path, file_metadata = None):
+    contents = os.listdir(current_path)
+
+    if file_metadata != None:
+        file_details = parse_file_metadata(current_path, file_metadata)
+    else:
+        file_details = gather_file_details(current_path, contents)
```

Comment:
Is the reason for this ordering having non-default behavior first? because to me, having `if ... != ...` feels weird, but I do understand your reasoning.

Reply:
Just a general pattern that I use of handling exceptional/edge/non-default cases first. This calls attention to the fact that they are exceptional and helps with preventing bugs.